src/main/scala/streamer/StreamerTop.scala

```diff
@@ -0,0 +1,119 @@
+package streamer
+
+import chisel3._
+import chisel3.util._
+
+/** This class represents the csr input and output ports of the streamer top
+  * module
+  *
+  * @param csrAddrWidth
+  *   csr registers address width
+  */
+class StreamerTopCsrIO(csrAddrWidth: Int) extends Bundle {
+
+  val req = Flipped(Decoupled(new CsrReq(csrAddrWidth)))
+  val rsp = Decoupled(new CsrRsp)
+
+}
+
+/** This class represents the input and output ports of the streamer top module
+  *
+  * @param params
+  *   the parameters class instantiation for the streamer top module
+  */
+class StreamerTopIO(
+    params: StreamerTopParams = new StreamerTopParams()
+) extends Bundle {
+
+  // ports for csr configuration
+  val csr = new StreamerTopCsrIO(params.csrAddrWidth)
+
+  // ports for data in and out
+  val data = new StreamerDataIO(
+    params
+  )
+}
+
+/** This class represents the streamer top module which adds the csr registers
+  * as well as csr read and write logic based on the streamer
+  *
+  * @param params
+  *   the parameters class instantiation for the streamer top module
+  */
+class StreamerTop(
+    params: StreamerTopParams = new StreamerTopParams()
+) extends Module
+    with RequireAsyncReset {
+  val io = IO(
+    new StreamerTopIO(
+      params
+    )
+  )
+
+  // csrManager instantiation
+  val csr_manager = Module(new CsrManager(params))
+
+  // streamer instantiation
+  val streamer = Module(
+    new Streamer(
+      params
+    )
+  )
+
+  // io.csr and csrManager input connection
+  csr_manager.io.csr_config_in <> io.csr
+
+  // below does streamer and csrManager output connection
+  // control signals
+  streamer.io.csr.valid := csr_manager.io.csr_config_out.valid
+  csr_manager.io.csr_config_out.ready := streamer.io.csr.ready
+
+  // splitting csrManager data ports to the streamer components
+  // temporal loop bounds
+  for (i <- 0 until params.temporalDim) {
+    streamer.io.csr.bits
+      .loopBounds_i(i) := csr_manager.io.csr_config_out.bits(i)
+  }
+
+  // temporal loop strides
+  for (i <- 0 until params.dataMoverNum) {
+    for (j <- 0 until params.temporalDim) {
+      streamer.io.csr.bits
+        .temporalStrides_csr_i(i)(j) := csr_manager.io.csr_config_out.bits(
+        params.temporalDim + i * params.temporalDim + j
+      )
+    }
+  }
+
+  // spatial loop strides
+  for (i <- 0 until params.spatialDim.length) {
+    for (j <- 0 until params.spatialDim(i)) {
+      streamer.io.csr.bits
+        .spatialStrides_csr_i(i)(j) := csr_manager.io.csr_config_out.bits(
+        params.temporalDim + params.dataMoverNum * params.temporalDim + i * params
+          .spatialDim(
+            i
+          ) + j
+      )
+    }
+  }
```

Comment:
can you write some comments/documentation how the final csr addresses will look like?

Reply:
> can you write some comments/documentation how the final csr addresses will look like?
Ok! More comments added. Maybe it is better if I also write it in README?

Reply:
> I also would like to know how many CSRs we are using right now to set all of these values.

See:   val csrNum: Int =
    temporalDim + dataMoverNum * temporalDim + spatialDim.sum + dataMoverNum + 1

> How many CSRs are required to control the spatial loop strides?

It is spatialDim.sum. It depends on the spatial dim related to the accelerator.

> IIUC right now it looks like all control is just passed to one giant flat signal: `csr_manager.io.csr_config_out.bits` I think it would be more understandable if we make this a bit more hierarchical maybe? What do others think?

Doesn't make much changes to me. I think both are ok!